File: src/thermo_receive.js

/**
 * Copyright 2017, Google, Inc.
 * Licensed under the Apache License, Version 2.0 (the `License`);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an `AS IS` BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var five = require('johnny-five');
var program = require('commander');

// https://www.npmjs.com/package/commander
program.description('The receiving end for the temperature/fan with device example')
  .option('--project_id <project_id>', 'GCP cloud project name. Only the name, not path.')
  .option('--device_port <device_port>', 'If device is plugged in, what port is it plugged into.')
  .option('--pubsub_topic <pubsub_topic>', 'Topic name from pubsub associated with the device registry.')
  .option('--pubsub_subscription <pubsub_subscription>', 'Subscription name from pubsub associated with the device regsitry.')
  .option('--service_account_json <service_account_json>', 'Path to the service account json file')
  .parse(process.argv);

// fail out if we don't have any of our required args.
if (!program.project_id) 
  throw new Error('--project_id required')
if (!program.device_port) 
  throw new Error('--device_port required')
if (!program.pubsub_topic)
  throw new Error('--pubsub_topic required')
if (!program.pubsub_subscription) 
  throw new Error('--pubsub_subscription required')
if (!program.service_account_json) 
  throw new Error('--service_account_json required')

// Information on creating service accounts and generating credentials:
// https://cloud.google.com/iam/docs/service-accounts?hl=en_US&_ga=1.221302744.1127203517.1497463231
var pubsub = require('@google-cloud/pubsub')({
  projectId: program.project_id,
  keyFilename: program.service_account_json
});
const pubsub_topic = `projects/${program.project_id}/topics/${program.pubsub_topic}`;
const pubsub_subscription = `projects/${program.project_id}/subscriptions/${program.pubsub_subscription}`;

// http://johnny-five.io/
var board = new five.Board({
    port: program.device_port
});

board.on("ready", function () {

    // https://googlecloudplatform.github.io/google-cloud-node/#/docs/pubsub/0.11.0/pubsub/subscription
    var subscription = pubsub.subscription(pubsub_subscription, {
      autoAck: true
    });

    subscription.on("message", function(message) {
      var led_70 = new five.Led(2);
      var led_73 = new five.Led(3);
      var led_76 = new five.Led(4);
      var led_79 = new five.Led(5);
      var led_82 = new five.Led(6);
      var led_85 = new five.Led(7);
      var led_88 = new five.Led(8);

      var msgJson = message.data;
      if (msgJson.temperature > 70.0) {
        led_70.on();
      }
      else {
        led_70.off();
      }

      if (msgJson.temperature > 73.0) {
        led_73.on();
      }
      else {
        led_73.off();
      }

      if (msgJson.temperature > 76.0) {
        led_76.on();
      }
      else {
        led_76.off();
      }

      if (msgJson.temperature > 79.0) {
        led_79.on();
      }
      else {
        led_79.off();
      }

      if (msgJson.temperature > 82.0) {
        led_82.on();
      }
      else {
        led_82.off();
      }

      if (msgJson.temperature > 85.0) {
        led_85.on();
      }
      else {
        led_85.off();
      }

      if (msgJson.temperature > 88.0) {
        led_88.on();
      }
      else {
        led_88.off();
      }

    });
});

// 
board.on("exit", function() {

    var led_70 = new five.Led(2);
    var led_73 = new five.Led(3);
    var led_76 = new five.Led(4);
    var led_79 = new five.Led(5);
    var led_82 = new five.Led(6);
    var led_85 = new five.Led(7);
    var led_88 = new five.Led(8);
  
    led_70.off();
    led_73.off();
    led_76.off();
    led_79.off();
    led_82.off();
    led_85.off();
    led_88.off();
});

File: src/thermo_receive.py
# Copyright 2017, Google, Inc.
# Licensed under the Apache License, Version 2.0 (the `License`);
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an `AS IS` BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# This script assumes that you've setup the Google Cloud SDK and init'd
# using the project the telemetry data is being sent to. The Service Account
# must have Pub/Sub read permissions only

import argparse
import rainbowhat as rh
import time
import json

from google.cloud import pubsub

led_0 = [0,255,200]
led_1 = [0,255,0]
led_2 = [100,255,0]
led_3 = [150,200,0]
led_4 = [200,100,0]
led_5 = [255,50,0]
led_6 = [255,0,0]


def receive_message(topic_name, subscription_name, service_account_file):
  pubsub_client = pubsub.Client.from_service_account_json(service_account_file)
  topic = pubsub_client.topic(topic_name)
  subscription = topic.subscription(subscription_name)

  results = subscription.pull(return_immediately=True)

  for ack_id, message in results:
    msg_json = json.loads(message.data)
    temp = msg_json['temperature']

    rh.display.clear()
    rh.display.print_float(temp)
    rh.display.show()

    if temp > 68:
      r,g,b = led_0
      rh.rainbow.set_pixel(0, r, g, b)
    if temp > 70:
      r,g,b = led_1
      rh.rainbow.set_pixel(1, r, g, b)
    if temp > 72:
      r,g,b = led_2
      rh.rainbow.set_pixel(2, r, g, b)
    if temp > 74:
      r,g,b = led_3
      rh.rainbow.set_pixel(3, r, g, b)
    if temp > 76:
      r,g,b = led_4
      rh.rainbow.set_pixel(4, r, g, b)
    if temp > 78:
      r,g,b = led_5
      rh.rainbow.set_pixel(5, r, g, b)
    if temp > 80:
      r,g,b = led_6
      rh.rainbow.set_pixel(6, r, g, b)
    rh.rainbow.show()

  if results:
    subscription.acknowledge([ack_id for ack_id, message in results])

def parse_command_line_args():
  parser = argparse.ArgumentParser(description='Raspberry Pi script to receieve temperature telemetry data from Pub/Sub')
  parser.add_argument(
    '--service_account_json_file', required=True, help='Full path and name of service account credentials file.')
  parser.add_argument(
    '--pubsub_topic', required=True, help='Cloud Pub/Sub topic name.')
  parser.add_argument(
    '--pubsub_subscription', required=True, help='Cloud Pub/Sub subscription name.')
  return parser.parse_args()

args = parse_command_line_args()

while True:
  receive_message(args.pubsub_topic, args.pubsub_subscription, args.service_account_json_file)
  time.sleep(0.5)
